dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -151,6 +151,48 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
     */
   def project[U](implicit projector: SmartProject[T,U]): TypedDataset[U] = projector.apply(this)
 
+  /** Returns a new [[TypedDataset]] that contains the elements of both this and the `other` [[TypedDataset]]
+    * combined.
+    *
+    * Note that, this function is not a typical set union operation, in that it does not eliminate
+    * duplicate items. As such, it is analogous to `UNION ALL` in SQL.
+    *
+    * Differs from `Dataset#union` by aligning fields if possible.
+    * It will not compile if `Datasets` have not compatible schema.
+    *
+    * Example:
+    * {{{
+    *   case class Foo(x: Int, y: Long)
+    *   case class Bar(y: Long, x: Int)
+    *   case class Faz(x: Int, y: Int, z: Int)
+    *
+    *   df1: TypedDataset[Foo] = ...
+    *   df2: TypedDataset[Bar] = ...
+    *   df3: TypedDataset[Faz] = ...
+    *
+    *   df1.union(df2): TypedDataset[Foo]
+    *
+    *   // fails to compile
+    *   df1.union(df3)
+    * }}}
+    *
+    * apache/spark
+    */
+  def union[U: TypedEncoder](other: TypedDataset[U])(implicit union: Union[T, U]): TypedDataset[T] =
```

Comment:
How about:
```scala
def union[U: TypedEncoder](other: TypedDataset[U])(implicit project: SmartProject[T, U]): TypedDataset[T] =  project(this).union(other)
```

Reply:
Yes, this works, just a bit different behavior but it seems like it is desired so I'm going with this :) One difference I did is returning type ```T``` (first dataset)
I'm using just ```union``` - should it be something different?